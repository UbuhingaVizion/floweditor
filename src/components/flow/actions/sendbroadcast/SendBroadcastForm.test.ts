import SendBroadcastForm from 'components/flow/actions/sendbroadcast/SendBroadcastForm';
import { ActionFormProps } from 'components/flow/props';
import { ComposeAttachment } from 'flowTypes';
import { composeComponentTestUtils, mock } from 'testUtils';
import { createBroadcastMsgAction, getActionFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const action = createBroadcastMsgAction();
const { setup } = composeComponentTestUtils<ActionFormProps>(
  SendBroadcastForm,
  getActionFormProps(action)
);

export const getTestAttachments = (numFiles = 2): ComposeAttachment[] => {
  const attachments = [];
  let index = 1;
  while (index <= numFiles) {
    const test = 'test' + index;
    const attachment = {
      uuid: test,
      content_type: 'image/png',
      type: 'image/png',
      filename: 'name_' + test,
      url: 'url_' + test,
      size: 1024,
      error: null
    } as ComposeAttachment;
    attachments.push(attachment);
    index++;
  }
  return attachments;
};

describe(SendBroadcastForm.name, () => {
  describe('render', () => {
    it('should render self, children with base props', () => {
      const { wrapper } = setup(true);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render an empty form with no action', () => {
      const { wrapper, instance } = setup(true, {
        $merge: {
          nodeSettings: { originalNode: null }
        }
      });

      expect(instance.state).toMatchSnapshot();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('handles recipent change', () => {
      const { instance } = setup(true, {
        $merge: { updateSendBroadcastForm: jest.fn().mockReturnValue(true) }
      });
      instance.handleRecipientsChanged([{ id: 'group-0', name: 'My Group' }]);
      expect(instance.state).toMatchSnapshot();
    });

    it('handles compose text change', () => {
      const { instance, props } = setup(true, {
        $merge: { updateSendBroadcastForm: jest.fn().mockReturnValue(true) }
      });
      const compose = JSON.stringify({ text: 'Some message', attachments: [] });
      instance.handleComposeChanged(compose);
      expect(instance.state).toMatchSnapshot();
    });

    it('handles compose attachments change', () => {
      const { instance, props } = setup(true, {
        $merge: { updateSendBroadcastForm: jest.fn().mockReturnValue(true) }
      });
      const text = 'Some message with an attachment';
      const attachments = getTestAttachments(1);
      const compose = JSON.stringify({ text: text, attachments: attachments });
      instance.handleComposeChanged(compose);
      expect(instance.state).toMatchSnapshot();
    });

    it('should allow switching from router', () => {
      const { instance, props } = setup(true, {
        $merge: { updateAction: jest.fn() },
        nodeSettings: { $merge: { originalAction: null } }
      });

      instance.handleRecipientsChanged([{ id: 'group-0', name: 'My Group' }]);
      const text = 'Some message with an attachment';
      const attachments = getTestAttachments(1);
      const compose = JSON.stringify({ text: text, attachments: attachments });
      instance.handleComposeChanged(compose);
      instance.handleSave();

      expect(props.updateAction).toMatchCallSnapshot();
    });
  });
});
