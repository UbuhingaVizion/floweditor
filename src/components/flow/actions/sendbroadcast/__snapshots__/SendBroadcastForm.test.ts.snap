// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendBroadcastForm render should render an empty form with no action 1`] = `
Object {
  "compose": Object {
    "value": "{\\"text\\":\\"\\",\\"attachments\\":[]}",
  },
  "recipients": Object {
    "value": Array [],
  },
  "valid": false,
}
`;

exports[`SendBroadcastForm render should render an empty form with no action 2`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="send_broadcast"
  title="Send Broadcast"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "component": [Function],
        "description": "Send somebody else a message",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "text",
        ],
        "name": "Send Broadcast",
        "type": "send_broadcast",
      }
    }
    onChange={[MockFunction]}
  />
  <AssetSelector
    assets={
      Object {
        "items": Object {},
        "type": "contact",
      }
    }
    entry={
      Object {
        "value": Array [],
      }
    }
    expressions={true}
    multi={true}
    name="Recipients"
    onChange={[Function]}
    placeholder="Select Contacts"
    searchable={true}
  />
  <p />
  <ComposeElement
    attachments={true}
    chatbox={true}
    counter={true}
    entry={
      Object {
        "value": "{\\"text\\":\\"\\",\\"attachments\\":[]}",
      }
    }
    name="Compose"
    onChange={[Function]}
  />
</Dialog>
`;

exports[`SendBroadcastForm render should render self, children with base props 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="send_broadcast"
  title="Send Broadcast"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "component": [Function],
        "description": "Send somebody else a message",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "text",
        ],
        "name": "Send Broadcast",
        "type": "send_broadcast",
      }
    }
    onChange={[MockFunction]}
  />
  <AssetSelector
    assets={
      Object {
        "items": Object {},
        "type": "contact",
      }
    }
    entry={
      Object {
        "value": Array [
          Object {
            "id": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Cat Fanciers",
            "type": "group",
          },
          Object {
            "id": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Cat Facts",
            "type": "group",
          },
          Object {
            "id": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "missing": false,
            "name": "Kellan Alexander",
            "type": "contact",
          },
          Object {
            "id": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "missing": false,
            "name": "Norbert Kwizera",
            "type": "contact",
          },
          Object {
            "id": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "missing": false,
            "name": "Rowan Seymour",
            "type": "contact",
          },
        ],
      }
    }
    expressions={true}
    multi={true}
    name="Recipients"
    onChange={[Function]}
    placeholder="Select Contacts"
    searchable={true}
  />
  <p />
  <ComposeElement
    attachments={true}
    chatbox={true}
    counter={true}
    entry={
      Object {
        "value": "{\\"text\\":\\"Hello World\\",\\"attachments\\":[]}",
      }
    }
    name="Compose"
    onChange={[Function]}
  />
</Dialog>
`;

exports[`SendBroadcastForm updates handles compose attachments change 1`] = `
Object {
  "compose": Object {
    "validationFailures": Array [],
    "value": "{\\"text\\":\\"Some message with an attachment\\",\\"attachments\\":[{\\"uuid\\":\\"test1\\",\\"content_type\\":\\"image/png\\",\\"type\\":\\"image/png\\",\\"filename\\":\\"name_test1\\",\\"url\\":\\"url_test1\\",\\"size\\":1024,\\"error\\":null}]}",
  },
  "recipients": Object {
    "value": Array [
      Object {
        "id": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        "name": "Cat Fanciers",
        "type": "group",
      },
      Object {
        "id": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        "name": "Cat Facts",
        "type": "group",
      },
      Object {
        "id": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        "missing": false,
        "name": "Kellan Alexander",
        "type": "contact",
      },
      Object {
        "id": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "missing": false,
        "name": "Norbert Kwizera",
        "type": "contact",
      },
      Object {
        "id": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        "missing": false,
        "name": "Rowan Seymour",
        "type": "contact",
      },
    ],
  },
  "valid": true,
}
`;

exports[`SendBroadcastForm updates handles compose text change 1`] = `
Object {
  "compose": Object {
    "validationFailures": Array [],
    "value": "{\\"text\\":\\"Some message\\",\\"attachments\\":[]}",
  },
  "recipients": Object {
    "value": Array [
      Object {
        "id": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        "name": "Cat Fanciers",
        "type": "group",
      },
      Object {
        "id": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        "name": "Cat Facts",
        "type": "group",
      },
      Object {
        "id": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        "missing": false,
        "name": "Kellan Alexander",
        "type": "contact",
      },
      Object {
        "id": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "missing": false,
        "name": "Norbert Kwizera",
        "type": "contact",
      },
      Object {
        "id": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        "missing": false,
        "name": "Rowan Seymour",
        "type": "contact",
      },
    ],
  },
  "valid": true,
}
`;

exports[`SendBroadcastForm updates handles recipent change 1`] = `
Object {
  "compose": Object {
    "value": "{\\"text\\":\\"Hello World\\",\\"attachments\\":[]}",
  },
  "recipients": Object {
    "validationFailures": Array [],
    "value": Array [
      Object {
        "id": "group-0",
        "name": "My Group",
      },
    ],
  },
  "valid": true,
}
`;

exports[`SendBroadcastForm updates should allow switching from router 1`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "content_type": "image/png",
        "error": null,
        "filename": "name_test1",
        "size": 1024,
        "type": "image/png",
        "url": "url_test1",
        "uuid": "test1",
      },
    ],
    "contacts": Array [],
    "groups": Array [],
    "legacy_vars": Array [],
    "text": "Some message with an attachment",
    "type": "send_broadcast",
    "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
  },
]
`;
